MDEV-35984 - Add MTR tests for MySQL (5.7) to MariaDB migration

PhysicsTing opened the pull request:
## Description

Currently upgrades from MySQL 5.7 to MariaDB fail on certain edge
cases. This commit adds MTR tests to reproduce those issues. The purpose
of this is to make the problems visible and facilitate the development
of the proper upgrade code in MariaDB. The MTR tests are added to skip
list as they will fail for all branches until these incompatibilities
are addressed.

Populated a MySQL 5.7 data directory from official MySQL docker image.
The scripts are stored in mysql-test/lib, and the data directory is
stored in std_data to be consumed by MTR directly.

Add MTR tests for the usual migration procedure from MySQL to MariaDB.
  1. Start MariaDB from the MySQL data directory
  2. Run mariadb-upgrade
  3. Verify that the upgrade succeed

Also highlight the incompatibilities that are either not documented
before or documented but the detailed scenario are not clarified. These
tests will fail for now but can be used as references for future to
improve these incompatibilities.

Focus on MySQL 5.7 migration for now.

## How can this PR be tested?

Newly created MTR tests failed but that is intentional.
Existing MTR tests succeed as expected.

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a bug fix and the PR is based against the earliest maintained branch in which the bug can be reproduced: 10.5*

## Backward compatibility
Backward compatibility is not compromised, and will be improved based on the newly added tests.

## Copyright
All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.

PhysicsTing committed:
Add MTR tests for MySQL (5.7) to MariaDB (10.5) migration

Currently upgrades from MySQL 5.7 to MariaDB 10.5 fail on certain edge
cases. This commit adds MTR tests to reproduce those issues. The purpose
of this is to make the problems visible and facilitate the development
of the proper upgrade code in MariaDB. The MTR tests are added to skip
list as they will fail for all branches until these incompatibilities
are addressed.

Populated a MySQL 5.7 data directory from official MySQL docker image.
The scripts are stored in mysql-test/lib, and the data directory is
stored in std_data to be consumed by MTR directly.

Add MTR tests for the usual migration procedure from MySQL to MariaDB.
  1. Start MariaDB from the MySQL data directory
  2. Run mariadb-upgrade
  3. Verify that the upgrade succeed

Also highlight the incompatibilities that are either not documented
before or documented but the detailed scenario are not clarified. These
tests will fail for now but can be used as references for future to
improve these incompatibilities.

Focus on MySQL 5.7 migration for now.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=2170) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=2170) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=2170) it.</sub>

grooverdan commented:
@PhysicsTing I suspect you're still after feedback on this approach.

We could also leave this to a non-mtr test in bb. Are you committed to a MTR approach?

dlenski commented:
> We could also leave this to a non-mtr test in bb. Are you committed to a MTR approach?

What's wrong with having this as an MTR test?

@PhysicsTing's approach here provides very concise documentation of all the known cases where a MySQL 5.7 data directory is incompatible with MariaDB, and will result in data loss/corruption if the engine is replaced with MariaDB.

In this form, it's also a roadmap and an automated framework for testing solutions to each of these incompatibilities.

For example, if a developer is trying to implement loading of InnoDB "native partitioning" tables in MariaDB, it can be quickly tested by uncommenting the `--error ER_NO_SUCH_TABLE_IN_ENGINE` here, and checking the .result for `select * from idb_partition_t1`:

https://github.com/MariaDB/server/blob/7bd148adb2706ffaec6fe5116478d8d4d1d3d6d6/mysql-test/suite/mysql_migration/mysql_migration.test#L65-L77

PhysicsTing commented:
Thank you both for the feedback and reply.

I used MTR mainly because at this point the actual test result matters less than highlighting to people these compatibilities in an explicit way (e.g. how to reproduce), and MTR framework empowers this well. So I agree with what @dlenski pointed out, but would love to hear if a non-MTR approach would potentially benefit more.

dr-m commented on mysql-test/suite/mysql_migration/mysql_migration.sql:
```diff
@@ -0,0 +1,108 @@
+
+-- Insert large data to make sure the compression will take place
+DELIMITER $$
+CREATE PROCEDURE populatePageCompTable()
```
It could be good to test that procedure definitions are the same after upgrading. I do not see the name of this procedure anywhere except this statement and the subsequent `call` statement.

dlenski replied:
> It could be good to test that procedure definitions are the same after upgrading.

In order to test that thoroughly, all different kinds of procedures and metadata (e.g. `SQL SECURITY DEFINER` vs. `… INVOKER`, `DETERMINISTIC`, etc.) would have to be created and verified.

That does seem like a good idea for coverage, but perhaps a bit outside the scope of this PR, which is focused on documenting the already-known cases of incompatibilities in MySQL 5.7 → MariaDB data directory migration.

dr-m replied:
Sure, there could be better frameworks for testing the compatibility of stored procedures. Perhaps something like the Random Query Generator could be employed to cover the entire MySQL 5.7 grammar for stored procedures, and perhaps we should also test that the execution is compatible. I agree that something like it is out of the scope of this PR.

Still, it might be better to not waste the opportunity to ensure that `SHOW PROCEDURE`  produces the expected output for these few special cases.

dr-m commented on mysql-test/suite/mysql_migration/mysql_migration.test:
```diff
@@ -0,0 +1,206 @@
+--echo #
+
+#
+# Because undo logs are implemented differently in MariDB and MySQL. This test is necessary to verify
+# that MariaDB undo logs works well with MySQL datadir
```
The problem is that MySQL 5.7 introduced an incompatibility with MySQL 5.6, which MariaDB avoided, by applying 124bae082bf17e9af1fc77f78bbebd019635be5c.

If something was upgraded from MySQL 5.6 or earlier (`innodb_rollback_segments=128` was the default already in 5.5) to 5.7 and then to MariaDB, then the rollback segments 1 to 32 could contain garbage and cause some trouble. I do not think that a test of creating a database in 5.7 and upgrading to MariaDB can reveal much trouble with regard to that.

In any case, it could be useful to cause MySQL 5.7 rollback segments 1 to 32 to be written by executing operations on tables created with `CREATE TEMPORARY TABLE`.

dr-m commented on mysql-test/suite/mysql_migration/mysql_migration.test:
```diff
@@ -0,0 +1,206 @@
+
+SHOW TABLES;
+
+--error ER_TABLE_CORRUPT # Allow the test to continue. Remove once this incompatibility is resolved
+select count(*) as EXPECT_3000 from idb_page_comp_t1;
```
Do we really want to support this file format? The way the page is compressed probably differs from both already existing implementations of `page_compressed` in MariaDB. (The second file format was introduced in 6b6fa3cdb16ae7b4bc9e307c7d9b9012a055548c.)

Just like with anything that uses `CREATE TABLESPACE`, the solution here could be to require such tables to be rebuilt before upgrading.

dlenski replied:
> Do we really want to support this file format?

> Do we really want to implement the MySQL 5.7 way of encrypting tables?

According to [the docs on upgrading from MySQL → MariaDB](https://mariadb.com/kb/en/upgrading-from-mysql-to-mariadb/):

“you can **in most cases** just uninstall MySQL [5.7] and install MariaDB and you are good to go. There is no need to dump and restore databases. As with any upgrade, we recommend making a backup of your data beforehand.”

The phrase “in most cases” is doing a lot of work there.

One of the purposes of this PR is to _clearly enumerate, document, and track_ the cases where this does not Just Work, but where data will be lost or corrupted in replacing MySQL 5.7 with MariaDB.

dr-m replied:
Yes, it is good to document the incompatibilities. I would suggest to write a longer explanatory comment why this particular incompatibility is not likely to be removed.

The MySQL 5.7 encryption and sparse-file compression are similar to what was implemented and released in MariaDB 10.1 implemented earlier. I do not think it makes sense to support either format. In fact, I agree that [the practical usability of sparse-file compression is questionable](https://smalldatum.blogspot.com/2015/10/wanted-file-system-on-which-innodb.html). If you want that type of compression, I believe that the best option is thinly provisioned computational storage, such as a ScaleFlux device on which we performed some testing. It won hands down, both in terms of speed and compression ratio.

Supporting the MySQL 5.7 encryption could require implementing some plugin interface adapter. I would be reluctant to do that.

Those (as well as the `CREATE TABLESPACE`) are the only things that I think are technically infeasible to support. There are other compatibility issues, such as [MDEV-29253](https://jira.mariadb.org/browse/MDEV-29253) for supporting the `.frm` format that is used by MySQL 5.7 partitioning, which are easier and more feasible to fix.

dr-m commented on mysql-test/suite/mysql_migration/mysql_migration.test:
```diff
@@ -0,0 +1,206 @@
+
+SHOW TABLES;
+
+--error ER_NO_SUCH_TABLE_IN_ENGINE # Allow the test to continue. Remove once this incompatibility is resolved
+select * from idb_encryption_t1;
```
Do we really want to implement the MySQL 5.7 way of encrypting tables? Such tables could be rebuilt before upgrading.

If MySQL 5.7 can combine encryption and compression, that should be tested as well. Likewise, encryption with `ROW_FORMAT=COMPRESSED` and `KEY_BLOCK_SIZE` of `1`, `2`, `4` needs to be tested.

dlenski replied:
Right now, many MySQL users are going to look at [the docs on upgrading from MySQL → MariaDB](https://mariadb.com/kb/en/upgrading-from-mysql-to-mariadb/) and see…

  * "you can **_in most cases_** just uninstall MySQL [5.7] and install MariaDB and you are good to go. There is no need to dump and restore databases. As with any upgrade, we recommend making a backup of your data beforehand."

If MySQL users have encrypted tables, this will be a bad surprise, because those tables _won't work_ after migrating to MariaDB.

> Such tables could be rebuilt before upgrading. Do we really want to implement the MySQL 5.7 way of encrypting tables? 

That I don't know, but…

Surely the answer **is not** to continue surprising MySQL users with "oops, your table is corrupt" after they attempt to migrate to MariaDB, right?

That kind of uncertainty is precisely what will discourage a lot of users from switching.

Another possibilities would be to give a substantially more informative error than `ER_NO_SUCH_TABLE_IN_ENGINE`. How about at least try to enumerate and explain the on-disk files? Perhaps something like `ER_INCOMPATIBLE_TABLE_FORMAT`:

  * "This table was produced by a version of MySQL which stores it in an on-disk format that this version of MariaDB does not directly support, because $REASON. See https://mariadb.com/kb/en/upgrading-from-mysql-to-mariadb#known-incompatibilities. Please revert to the old engine version and try converting the table to a supported format before migrating to MariaDB."

dr-m reviewed (commented):
I think that this needs to cover every `innodb_page_size` as well.

ottok commented:
#2555 fixes at least on testable MySQL 5.7 to MariaDB upgrade issue. This PR should be finalized so we can catch all of those issues.

PhysicsTing commented:
Previous PR is in draft state and therefore is not a good state for discussion. Refresh this PR after a long while. The changes are:
- Instead of spinning up a docker during MTR which won't work in most of CI system, use a pre-generated data directory.
- Instead of expect failures in MTR test, add the MTR to skip list.
- Rebase on 10.5 so the improvement of incompatibilities can be observed as version progresses.

PhysicsTing commented:
Pushed again. In the latest update I addressed comments from [Marko Mäkelä](https://github.com/dr-m):

- Tested that the specific stored procedure does not change upgrade migration. I did not cover other types of stored procedure which I think is outside of the scope and requires more knowledge on the topic. That can be expanded later.
- Created temp table in MySQL 5.7 to fill in the first 32 undo segments. Commented that this does not cover cases 5.6 -> 5.7 -> MariaDB.
- Explained that the MySQL compression format is unlikely to be supported in MariaDB.

dr-m reviewed (commented):
I think that this test could serve a useful purpose of documenting the known incompatibilites, as well as ensuring that no further incompatibilities will be introduced.

Please omit the files `ib_logfile0` and `ib_logfile1`. They are logically empty (there must have been a clean shutdown of the MySQL 5.7 server) and therefore they just waste space. We have log file format upgrade tests in more compact (and editable) form, in the tests `innodb.log_corruption` and `encryption.innodb_encrypt_log_corruption`. If the log files are missing, the upgrade must work based on the `FIL_PAGE_FILE_FLUSH_LSN` that is stored in the `ibdata1` file. This is something that was improved later in [MDEV-27199](https://jira.mariadb.org/browse/MDEV-27199), in a compatible way that allows MariaDB server to upgrade from older data files where the log files have been deleted.

In `ibdata1` you could zero out the doublewrite buffer (pages 64 to 191) to make it better compressible. The following should not return any garbage, like it currently does:
```sh
od -Ax -t x1 -j 0x100000 -N 0x200000 mysql-test/std_data/mysql_upgrade/mysql/ibdata1
```


There are some other files that certainly do not matter; `ib_buffer_pool` is one of them, and possibly the `*.pem` files as well. I would suggest to strip down the copy of the data directory to the minimum to highlight the problems. Maybe you could include a script that can be run against MySQL 5.7 to create the files and to strip them down to the minimum? That would help in extending the test in the future, if needed. 
The compressible size of the `ibdata1` could possibly be reduced with some further tricks, such as executing a shutdown with `innodb_fast_shutdown=0` and subsequently zeroing out all pages that are not marked as allocated within the tablespace.

It could be a good idea to attempt `ALTER TABLE…IMPORT TABLESPACE` on all of the data files as well, when starting InnoDB on a normal MariaDB data directory, to see if all files are correctly flagged as incompatible.

I would suggest at least the following change to the test, as well as some adjustment of the result, to make the test actually pass, so that it can serve the two purposes that I mentioned:
```diff
diff --git a/mysql-test/suite/mysql_migration/mysql_migration.test b/mysql-test/suite/mysql_migration/mysql_migration.test
index cd049e56665..085ad0fe194 100644
--- a/mysql-test/suite/mysql_migration/mysql_migration.test
+++ b/mysql-test/suite/mysql_migration/mysql_migration.test
@@ -54,7 +54,7 @@ let $restart_parameters=--datadir=$MYSQLTEST_VARDIR/tmp/mysql;
 --echo #
 
 #
-# MariDB doc:
+# MariaDB doc:
 #   MariaDB 10.3 does not support MySQL 5.7's “native” InnoDB partitioning handler.
 #
 # Please make sure the coresponding data exists in MySQL datadir via $POPULATE_MYSQL_DATADIR script
@@ -88,6 +88,7 @@ use idb_page_comp_db1;
 
 SHOW TABLES;
 
+--error ER_NO_SUCH_TABLE_IN_ENGINE
 select count(*) as EXPECT_3000 from idb_page_comp_t1;
 
 --echo #
@@ -102,7 +103,7 @@ SHOW CREATE PROCEDURE idb_page_comp_db1.populatePageCompTable;
 --echo #
 
 #
-# MariDB doc:
+# MariaDB doc:
 #   MariaDB 10.3 implements InnoDB encryption in a different way to MySQL 5.7
 #
 # Please make sure the corresponding data exists in MySQL datadir via $POPULATE_MYSQL_DATADIR script
@@ -112,6 +113,7 @@ use idb_encryption_db1;
 
 SHOW TABLES;
 
+--error ER_NO_SUCH_TABLE_IN_ENGINE
 select * from idb_encryption_t1;
 
 
@@ -133,6 +135,7 @@ use idb_tablespace_db1;
 
 SHOW TABLES;
 
+--error ER_NO_SUCH_TABLE_IN_ENGINE
 select * from idb_tablespace_t1;
 
 
@@ -141,7 +144,7 @@ select * from idb_tablespace_t1;
 --echo #
 
 #
-# Because undo logs are implemented differently in MariDB and MySQL. This test is necessary to verify
+# Because undo logs are implemented differently in MariaDB and MySQL. This test is necessary to verify
 # that MariaDB undo logs works well with MySQL datadir
 #
 
```
I tested this on 10.6 bb2e125d076d9475cbe04fc8c24fbaee39d76bda.

dr-m commented:
@PhysicsTing, it would be nice if you could update this to address my review comments.

gkodinov reviewed (changes requested):
This is a preliminary review. I sincerely hope you're still willing to work on this after all that time. In case you are, please:
 * rebase on the latest 10.11: this is the current branch now.
 * review and address Marko's comments
 * make sure the CLA buildbot is green. 
 * Watch and make sure that the buildbot tests are green once you do all of the above.

gkodinov commented:
There was no reply to my question from last month. Converting to draft. Please re-open when/if you resume working on this.
